relay-server/src/actors/spooler/mod.rs

```diff
@@ -179,148 +176,105 @@ impl FromMessage<RemoveMany> for Buffer {
+                (k.len(), k.into_iter().map(|k| k.estimated_size()).sum())
+            });
+            count += current_count;
+            self.used_memory -= current_size;
+        }
+        relay_statsd::metric!(
+            histogram(RelayHistograms::BufferEnvelopesMemory) = self.used_memory as f64
+        );
+
+        count
+    }
 
-            service.spool_config = Some(spool_config);
+    /// Dequeues the envelopes from the in-memory buffer and send them to provided `sender`.
+    fn dequeue(&mut self, keys: &Vec<QueueKey>, sender: mpsc::UnboundedSender<ManagedEnvelope>) {
+        for key in keys {
+            for value in self.buffer.remove(key).unwrap_or_default() {
+                self.used_memory -= value.estimated_size();
+                sender.send(value).ok();
+            }
         }
+        relay_statsd::metric!(
+            histogram(RelayHistograms::BufferEnvelopesMemory) = self.used_memory as f64
+        );
+    }
 
-        Ok(service)
+    /// Enqueues the envelope into in-memory buffer.
+    fn enqueue(&mut self, key: QueueKey, managed_envelope: ManagedEnvelope) {
+        self.used_memory += managed_envelope.estimated_size();
+        self.buffer.entry(key).or_default().push(managed_envelope);
+        relay_statsd::metric!(
+            histogram(RelayHistograms::BufferEnvelopesMemory) = self.used_memory as f64
+        );
     }
 
-    /// Estimates the db size by multiplying `page_count * page_size`.
-    async fn estimate_buffer_size(db: &Pool<Sqlite>) -> Result<i64, BufferError> {
-        let size: i64 = sql::current_size()
-            .fetch_one(db)
-            .await
-            .and_then(|r| r.try_get(0))
-            .map_err(BufferError::FileSizeReadFailed)?;
+    /// Returns `true` if the in-memory buffer is full, `false` otherwise.
+    fn is_full(&self) -> bool {
+        self.used_memory >= self.max_memory_size
+    }
+}
 
-        relay_statsd::metric!(histogram(RelayHistograms::BufferDiskSize) = size as u64);
-        Ok(size)
+/// The configuration which describes the on-disk [`BufferState`].
+#[derive(Debug)]
+struct OnDisk {
+    db: Pool<Sqlite>,
+    buffer_guard: Arc<BufferGuard>,
+    max_disk_size: usize,
+}
+
+impl OnDisk {
+    /// Creates a new [`OnDisk`] state.
+    fn new(db: Pool<Sqlite>, buffer_guard: Arc<BufferGuard>, max_disk_size: usize) -> Self {
+        Self {
+            db,
+            buffer_guard,
+            max_disk_size,
+        }
     }
 
     /// Saves the provided buffer to the disk.
     ///
     /// Returns an error if the spooling failed, and the number of spooled envelopes on success.
-    async fn do_spool(
-        db: &Pool<Sqlite>,
-        buffer: BTreeMap<QueueKey, Vec<ManagedEnvelope>>,
+    async fn spool(
+        &self,
+        buffer: &mut BTreeMap<QueueKey, Vec<ManagedEnvelope>>,
+        used_memory: &mut usize,
     ) -> Result<(), BufferError> {
+        let buffer = std::mem::take(buffer);
+        *used_memory = 0;
```

Comment:
Should we check if there's available space on the disk? What if there's only space for the first item in the buffer? We're not checking that.

Reply:
I can switch the condition , to check before `enqueue`, right now it's another way around. 
But also in case of errors while spooling we will lose the entire buffer or the items which cannot be spooled

Reply:
Isn't `used_memory` dropped shortly after calling `spool`, because after `spool` we transition into disk mode?